builtin/providers/aws/resource_aws_vpn_gateway.go

```diff
@@ -0,0 +1,328 @@
+        return err
+    }
+
+    log.Printf("[INFO] Deleting VPN gateway: %s", d.Id())
+
+    return resource.Retry(5*time.Minute, func() error {
+        err := ec2conn.DeleteVPNGateway(&ec2.DeleteVPNGatewayRequest{
+            VPNGatewayID: aws.String(d.Id()),
+        })
+        if err == nil {
+            return nil
+        }
+
+        ec2err, ok := err.(*aws.APIError)
+        if !ok {
+            return err
+        }
+
+        switch ec2err.Code {
+        case "InvalidVpnGatewayID.NotFound":
+            return nil
+        case "DependencyViolation":
+            return err // retry
+        }
+
+        return resource.RetryError{Err: err}
+    })
+}
+
+func resourceAwsVpnGatewayAttach(d *schema.ResourceData, meta interface{}) error {
+    ec2conn := meta.(*AWSClient).awsEC2conn
+
+    if d.Get("vpc_id").(string) == "" {
+        log.Printf(
+            "[DEBUG] Not attaching VPN Gateway '%s' as no VPC ID is set",
+            d.Id())
+        return nil
+    }
+
+    log.Printf(
+        "[INFO] Attaching VPN Gateway '%s' to VPC '%s'",
+        d.Id(),
+        d.Get("vpc_id").(string))
+
+    _, err := ec2conn.AttachVPNGateway(&ec2.AttachVPNGatewayRequest{
+        VPNGatewayID: aws.String(d.Id()),
+        VPCID:        aws.String(d.Get("vpc_id").(string)),
+    })
+    if err != nil {
+        return err
+    }
+
+	// A note on the states below: the AWS docs (as of July, 2014) say
+	// that the states would be: attached, attaching, detached, detaching,
+	// but when running, I noticed that the state is usually "available" when
+	// it is attached.
+
+	// Wait for it to be fully attached before continuing
+    log.Printf("[DEBUG] Waiting for VPN gateway (%s) to attach", d.Id())
+    stateConf := &resource.StateChangeConf{
+        Pending: []string{"detached", "attaching"},
+        Target:  "available",
+        Refresh: VpnGatewayAttachStateRefreshFunc(ec2conn, d.Id(), "available"),
+        Timeout: 1 * time.Minute,
+    }
+    if _, err := stateConf.WaitForState(); err != nil {
+        return fmt.Errorf(
+            "Error waiting for VPN gateway (%s) to attach: %s",
+            d.Id(), err)
+    }
+
+    return nil
+}
+
+func resourceAwsVpnGatewayDetach(d *schema.ResourceData, meta interface{}) error {
+    ec2conn := meta.(*AWSClient).awsEC2conn
+
+    // Get the old VPC ID to detach from
+    vpcID, _ := d.GetChange("vpc_id")
+
+    if vpcID.(string) == "" {
+        log.Printf(
+            "[DEBUG] Not detaching VPN Gateway '%s' as no VPC ID is set",
+            d.Id())
+        return nil
+    }
+
+    log.Printf(
+        "[INFO] Detaching VPN Gateway '%s' from VPC '%s'",
+        d.Id(),
+        vpcID.(string))
+
+    wait := true
+    err := ec2conn.DetachVPNGateway(&ec2.DetachVPNGatewayRequest{
+        VPNGatewayID: aws.String(d.Id()),
+        VPCID:        aws.String(d.Get("vpc_id").(string)),
+    })
+    if err != nil {
+        ec2err, ok := err.(*aws.APIError)
```

Comment:
If I'm not mistaken, the return here should be `aws.APIError `. Does the `*` work?